cmd/topology/topology_test.go

```diff
@@ -0,0 +1,248 @@
+	}
+	return nr.configPushErr
+}
+
+type resetable struct {
+	*notResetable
+	resetErr error
+}
+
+func (r *resetable) ResetCfg(ctx context.Context, ni node.Interface) error {
+	return r.resetErr
+}
+
+func NewR(pb *topopb.Node) (node.Implementation, error) {
+	return &resetable{&notResetable{pb: pb}, nil}, nil
+}
+
+func writeTopology(t *testing.T, topo *topopb.Topology) (*os.File, func()) {
+	t.Helper()
+	f, err := ioutil.TempFile("", "reset")
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer os.Remove(t.Name())
+	b, err := prototext.Marshal(topo)
+	if err != nil {
+		t.Fatalf("failed to marshal topology")
+	}
+	if _, err := f.Write(b); err != nil {
+		t.Fatalf("failed to write topology")
+	}
+	return f, func() { os.Remove(f.Name()) }
+}
+
+func TestReset(t *testing.T) {
+	confFile, err := ioutil.TempFile("", "reset")
+	if err != nil {
+		log.Fatal(err)
+	}
+	fmt.Fprintln(confFile, "some bytes")
+	defer os.Remove(confFile.Name())
+	tInstance := &topopb.Topology{
+		Nodes: []*topopb.Node{{
+			Name: "resettable1",
+			Type: topopb.Node_Type(1001),
+		}, {
+			Name: "resettable2",
+			Type: topopb.Node_Type(1001),
+		}, {
+			Name: "notresettable1",
+			Type: topopb.Node_Type(1002),
+		}},
+	}
+	fNoConfig, closer := writeTopology(t, tInstance)
+	defer closer()
+	tWithConfig := &topopb.Topology{
+		Nodes: []*topopb.Node{{
+			Name: "resettable1",
+			Type: topopb.Node_Type(1001),
+			Config: &topopb.Config{
+				ConfigData: &topopb.Config_Data{
+					Data: []byte("somebytes"),
+				},
+			},
+		}, {
+			Name: "resettable2",
+			Type: topopb.Node_Type(1001),
+			Config: &topopb.Config{
+				ConfigData: &topopb.Config_File{
+					File: confFile.Name(),
+				},
+			},
+		}, {
+			Name: "notresettable1",
+			Type: topopb.Node_Type(1002),
+		}},
+	}
+	fConfig, closer := writeTopology(t, tWithConfig)
+	defer closer()
+	tWithConfigRelative := &topopb.Topology{
+		Nodes: []*topopb.Node{{
+			Name: "resettable1",
+			Type: topopb.Node_Type(1001),
+			Config: &topopb.Config{
+				ConfigData: &topopb.Config_Data{
+					Data: []byte("somebytes"),
+				},
+			},
+		}, {
+			Name: "resettable2",
+			Type: topopb.Node_Type(1001),
+			Config: &topopb.Config{
+				ConfigData: &topopb.Config_File{
+					File: filepath.Base(confFile.Name()),
+				},
+			},
+		}, {
+			Name: "notresettable1",
+			Type: topopb.Node_Type(1002),
+		}},
+	}
+	fConfigRelative, closer := writeTopology(t, tWithConfigRelative)
+	defer closer()
+	tWithConfigDNE := &topopb.Topology{
+		Nodes: []*topopb.Node{{
+			Name: "resettable1",
+			Type: topopb.Node_Type(1001),
+			Config: &topopb.Config{
+				ConfigData: &topopb.Config_Data{
+					Data: []byte("somebytes"),
+				},
+			},
+		}, {
+			Name: "resettable2",
+			Type: topopb.Node_Type(1001),
+			Config: &topopb.Config{
+				ConfigData: &topopb.Config_File{
+					File: "dne",
+				},
+			},
+		}, {
+			Name: "notresettable1",
+			Type: topopb.Node_Type(1002),
+		}},
+	}
+	fConfigDNE, closer := writeTopology(t, tWithConfigDNE)
+	defer closer()
```

Comment:
Do these defers properly clean all the files up? Not sure if there are any issues defer a func and then overwriting the func and deferring it again. Probably not but want to make sure this is working as intended

Reply:
i am not overwriting anything just reuse a pointer and call it - the fact that the address changes doesn't change anything